File: day-03/rust/src/main.rs
fn part1(input: &str) -> u32 {
    let mut total = 0;

    let rows: Vec<&str> = input.lines().collect();

    let height = rows.len() as u32;
    let width = rows[0].len() as u32;

    for (y , row) in rows.iter().enumerate() {
        let y = y as u32;

        let line: Vec<char> = row.chars().collect();

        let mut start = None;
        let mut end = None;
        let mut count = 0;

        for x in 0..width {
            let digit = line[x as usize].to_digit(10);

            if let Some(digit) = digit {
                if start.is_none() {
                    start = Some(x);
                }

                end = Some(x);
                count = count * 10 + digit;
            } 
            
            if digit.is_none() || x == width - 1 {
                if let Some(start) = start {
                    if let Some(end) = end {
                        let x1 = if start > 0 { start - 1 } else { start };
                        let x2 = if end < width - 1 { end + 1 } else { end };

                        let y1 = if y > 0 { y - 1 } else { y };
                        let y2 = if y < height - 1 { y + 1 } else { y };

                        let mut valid = false;

                        'out: for y in y1..=y2 {
                            for x in x1..=x2 {
                                let char = rows[y as usize].chars().nth(x as usize).unwrap();
                                if char != '.' && !char.is_ascii_digit() {
                                    valid = true;
                                    break 'out;
                                }
                            }
                        }

                        if valid {
                            total += count;
                        }
                    }
                }

                start = None;
                end = None;
                count = 0;
            }
        }
    }

    total
}

fn part2(input: &str) -> u32 {
    let mut total = 0;

    let rows: Vec<&str> = input.lines().collect();

    let height = rows.len() as u32;
    let width = rows[0].len() as u32;

    let mut gears = Vec::new();

    for (y , row) in rows.iter().enumerate() {
        let y = y as u32;
        let line: Vec<char> = row.chars().collect();

        let mut start = None;
        let mut end = None;
        let mut count = 0;

        for x in 0..width {
            let digit = line[x as usize].to_digit(10);

            if let Some(digit) = digit {
                if start.is_none() {
                    start = Some(x);
                }

                end = Some(x);
                count = count * 10 + digit;
            } 
            
            if digit.is_none() || x == width - 1 {
                if let Some(start) = start {
                    if let Some(end) = end {
                        let x1 = if start > 0 { start - 1 } else { start };
                        let x2 = if end < width - 1 { end + 1 } else { end };

                        let y1 = if y > 0 { y - 1 } else { y };
                        let y2 = if y < height - 1 { y + 1 } else { y };

                        for y in y1..=y2 {
                            for x in x1..=x2 {
                                let char = rows[y as usize].chars().nth(x as usize).unwrap();
                                if char == '*' {
                                    if let Some((_, _, other)) = gears.iter().find(|(ox, oy, _)| { *ox == x && *oy == y}) {
                                        total += count * other;
                                    } else {
                                        gears.push((x, y, count));
                                    };

                                    // possible optimization: remove gears that are no longer valid
                                    // gears = gears.into_iter().filter(|(ox, oy, _)| { *ox != x || *oy != y}).collect();
                                }
                            }
                        }
                    }
                }

                start = None;
                end = None;
                count = 0;
            }
        }
    }

    total
}

fn main() {
    let input = include_str!("../../input.txt");

    println!("Part 1: {}", part1(input));
    println!("Part 2: {}", part2(input));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part1() {
        assert_eq!(
            part1(
                "467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598.."
            ),
            4361
        );
    }

    #[test]
    fn test_part2() {
        assert_eq!(
            part2(
                "467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598.."
            ),
            467835
        );
    }
}

File: day-03/c/src/main.c
#include <stdarg.h>

#define FC_IMPLEMENTATION
#include "../../../lib/c/fc.h"

#define GET_INPUT_IMPLEMENTATION
#include "../../../lib/c/get_input.h"

int part1(SV input) {
    int total = 0;

    SV clone = sv_clone(input);
    usize width = sv_split(&clone, '\n').count;
    usize height = (input.count + 1) / (width + 1); // +1 to account for newlines

    SV_FOR_EACH_SPLIT(y, line, NL_SV, &input) {

        usize start = -1;
        usize end = -1;
        usize count = 0;

        SV_FOR_EACH_SPLIT(x, data, EMPTY_SV, &line) {
            char c = data.data[0];

            if (isdigit(c)) {
                if (start == (usize)-1) {
                    start = x;
                }

                end = x;
                count = count * 10 + (c - '0');
            }

            if (!isdigit(c) || x == width - 1) {
                if (start != (usize)-1) {
                    usize x1 =  start > 0 ? start - 1 : start;
                    usize x2 = end < width - 1 ? end + 1 : end;

                    usize y1 = y > 0 ? y - 1 : y;
                    usize y2 = y < height - 1 ? y + 1 : y;

                    bool valid = false;

                    for (usize y = y1; y <= y2; y++) {
                        for (usize x = x1; x <= x2; x++) {
                            char test = input.data[y * (width + 1) + x];

                            if (test != '.' && !isdigit(test)) {
                                valid = true;
                                goto out;
                            }
                        }
                    }

                    out:

                    if (valid) {
                        // printf("%d", count);
                        total += count;
                    }
                }

                start = -1;
                end = -1;
                count = 0;
            }
        }
    }

    return total;
}

typedef struct {
    usize x;
    usize y;
    usize value;
} Element;

da_typedef(Element, Elements);

int part2(SV input) {
    int total = 0;

    SV clone = sv_clone(input);
    usize width = sv_split(&clone, '\n').count;
    usize height = (input.count + 1) / (width + 1); // +1 to account for newlines

    Elements elements = {0};

    SV_FOR_EACH_SPLIT(y, line, NL_SV, &input) {

        usize start = -1;
        usize end = -1;
        usize count = 0;

        SV_FOR_EACH_SPLIT(x, data, EMPTY_SV, &line) {
            char c = data.data[0];

            if (isdigit(c)) {
                if (start == (usize)-1) {
                    start = x;
                }

                end = x;
                count = count * 10 + (c - '0');
            }

            if (!isdigit(c) || x == width - 1) {
                if (start != (usize)-1) {
                    usize x1 =  start > 0 ? start - 1 : start;
                    usize x2 = end < width - 1 ? end + 1 : end;

                    usize y1 = y > 0 ? y - 1 : y;
                    usize y2 = y < height - 1 ? y + 1 : y;

                    for (usize y = y1; y <= y2; y++) {
                        for (usize x = x1; x <= x2; x++) {
                            char test = input.data[y * (width + 1) + x];

                            if (test == '*') {
                                for (usize i = 0; i < elements.count; i++) {
                                    Element* element = &elements.items[i];

                                    if (element->x == x && element->y == y) {
                                        total += element->value * count;
                                        goto out;
                                    }
                                }

                                Element element = {x, y, count};
                                da_push(&elements, element);
                            }
                        }
                    }
                }

                out:

                start = -1;
                end = -1;
                count = 0;
            }
        }
    }

    return total;
}

void test();
int main(int argc, char **argv) {
    if (argc > 1 && strcmp(argv[1], "--test") == 0) {
        test();
        return 0;
    }

    char* input = get_input(argc, argv);

    printf("Part 1: %d\n", part1(sv_from_cstr(input)));
    printf("Part 2: %d\n", part2(sv_from_cstr(input)));
}

void test() {
    printf("Testing part 1: ");
    fflush(stdout);

    ASSERT_EQ(int, part1(sv_from_cstr("467..114..\n\
...*......\n\
..35..633.\n\
......#...\n\
617*......\n\
.....+.58.\n\
..592.....\n\
......755.\n\
...$.*....\n\
.664.598..")), 4361, %d, %d);

    printf("PASSED\nTesting part 2: ");
    fflush(stdout);

    ASSERT_EQ(int, part2(sv_from_cstr("467..114..\n\
...*......\n\
..35..633.\n\
......#...\n\
617*......\n\
.....+.58.\n\
..592.....\n\
......755.\n\
...$.*....\n\
.664.598..")), 467835, %d, %d);


    printf("PASSED\n\nBoth tests passed!\n");
}